### ISSUE TITLE
wxdemo.py does not exit

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3948]
|<img alt="ciphergoth's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ciphergoth reported|
|-|-|
|Trac ID|trac#3948|
|Type|defect|
|Created|2009-08-01 11:33:05Z|
|Branch|https://github.com/twisted/twisted/tree/wxreactor-stop-fix-3948|

On Ubuntu "Jaunty Jackalope" i386, using the system packages for wxPython 2.8.9.1 and Twisted 8.2.0, I run 

python /usr/share/doc/twisted-doc/examples/wxdemo.py

The demo window appears, but clicking on "File > Exit" or the window manager close button has no effect.  Expected behaviour: the process should terminate.

I'm having trouble getting my own wxreactor-based application to exit properly, so a working demo that shows the proper way to do it would be much appreciated - thanks!

Attachments:

* [issue_3948.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1e8/1e8d0db4d285c13611cf3affd3dc3119d6569574/b25c9845b15b316aa0b59218935dac681b58b647.patch) (1276 bytes) - added by psykidellic on 2011-02-09 20:44:49Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3948 3948
type__defect defect
reporter__ciphergoth ciphergoth
priority__normal normal
milestone__ 
branch__branches_wxreactor_stop_fix_3948 branches/wxreactor-stop-fix-3948
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1249126385000000 1249126385000000
changetime__1297649534000000 1297649534000000
version__None None
owner__ 
cc__itamar cc__khorn cc__jesstess cc__thijs cc__psykidellic
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ciphergoth's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ciphergoth removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess commented|
|-|-|

I had to add self.Destroy() before reactor.stop() in DoExit() to make the window close on clicking "File > Exit", fwiw, but the window manager close button still doesn't work and the process still doesn't terminate either way.

**Comment by twisted-trac** [comment]
|<img alt="michael-fig's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|michael-fig commented|
|-|-|

I've confirmed this under hardy heron (Ubuntu 8.04) with Twisted-9.0.0 compiled from sources.

Even with a Destroy() added before the reactor.stop(), the process hangs (one thread is still active) and cannot be killed except with kill -9.

I am happy to help debug this, as it is a showstopper that will prevent me from using Twisted on my next project.

**Comment by twisted-trac** [comment]
|<img alt="michael-fig's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|michael-fig commented|
|-|-|

I think I found it...

The wxreactor was just calling _threadselect.stop directly, when it needed to post an event to its queue and wake it instead.  This seems to solve everything (I will test more thoroughly later, but for now it gets the wxdemo working without any trouble).

Here's the replacement wxreactor.py WxReactor.stop():
```
    def stop(self):
        """
        Stop the reactor.
        """
        if self._stopping:
            return
        self._stopping = True
        self.callFromThread(_threadedselect.ThreadedSelectReactor.stop, self)
```

**Comment by twisted-trac** [comment]
|<img alt="michael-fig's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|michael-fig commented|
|-|-|

Replying to [michael-fig](#note_4):
> I think I found it...

I would also suggest adding a note to the Twisted GUI documents: if you want to show a wx dialog from within a reactor callback, you should use wx.CallLater(myDialog.ShowModal), otherwise you'll hang the reactor until the reactor callback returns.

After discovering this tip, I'm pleased to say that the wxreactor works as expected and I'll be using it after all!

**Comment by twisted-trac** [comment]
|<img alt="michael-fig's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|michael-fig commented|
|-|-|

Here is code to monkey-patch the wxreactor if you don't have the ability to change the sources.  It uses a more correct technique, in which the stop method itself is called from the main thread:

```
from twisted.internet import wxreactor
wxreactor.install()
from twisted.internet import reactor

# Monkey-patch the wxreactor to make it exit correctly.
if wxreactor.WxReactor.callFromThread is not None:
    oldStop = wxreactor.WxReactor.stop
    def stopFromThread(self):
        self.callFromThread(oldStop, self)
    wxreactor.WxReactor.stop = stopFromThread
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

We had another person with same problem, and the patch fixed things for them. Moving this to review, since there's a patch and it seems to work.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|Screwtape commented|
|-|-|

I'm confused. There's no patch attached to this ticket, and no branch. Exactly what needs to be reviewed?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

Comment #4 is almost a patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to michael-fig|
|-|-|

I can't reproduce the working behavior reported by making the change in comment 4.  I also thought about the change for a while before testing it, and I could not think of a reason it would make a difference.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|psykidellic removed owner|
|-|-|

Find attached a complete patch - fix from comment 4 and top files - that fixes this issue. Using wx 2.8.10 on Ubuntu.


Basically, we need to stop the threaded wxreactor.stop from the main reactor thread rather than itself.


I really dont know how to write a test case for it but our internal GUI app (which is considerably complex as it uses bunch of custom controls) seems to work.


So if you have any suggestion on how to create one, I can do that - otherwise I really dont know how to write a testcase for this.


Also, lot of methods didnt have the 2 line breaks as mentioned in the styling guide, so I updated the same with this.


This patch applies to trunk at 30649.

**Comment by twisted-trac** [comment]
|<img alt="psykidellic's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|psykidellic commented|
|-|-|

FWIW, the problem is there on both Linux and Mac OS X using wx 2.8.10. And the changes fixes both the environment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [30728]) Branching to 'wxreactor-stop-fix-3948'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [30729]) Make sure threadedselectreactor.stop() is noticed by waking up all the reactor threads when that method is called.

refs [#3948](https://github.com/twisted/twisted/issues/3948)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [30731]) news fragment

refs [#3948](https://github.com/twisted/twisted/issues/3948)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_19"></a>|@itamarst set owner to @exarkun|
|-|-|

wxdemo.py now exits correctly - merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set status to `closed`|
|-|-|

(In [30740]) Merge wxreactor-stop-fix-3948

Author: exarkun
Reviewer: itamar
Fixes: [#3948](https://github.com/twisted/twisted/issues/3948)


Wake up the necessary threads in wxreactor when the reactor is stopped to ensure
the reactor stops in a timely manner.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
